File: glasses/src/sessions.test.ts
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { __setDshEnabled, flattenSessions, glyph, isDsh, liveState } from "./sessions.ts";
import type { AgentInfo, LiveSignals, SessionInfo } from "./types.ts";

function signals(overrides: Partial<LiveSignals> = {}): LiveSignals {
  return {
    bridgeAttached: true,
    transcriptAgeSec: null,
    lastRole: null,
    lastHasToolUse: false,
    question: null,
    questionOptions: [],
    tail: [],
    newPrUrls: [],
    ...overrides,
  };
}

function session(overrides: Partial<SessionInfo> = {}): SessionInfo {
  return {
    id: "s1",
    repo: "myrepo",
    status: "running",
    session: null,
    ...overrides,
  };
}

describe("liveState", () => {
  // XERK-235: two rules the web applies and this mirror did not. paneBusy is a
  // value on the record the host last pushed, so a host that dies mid-turn left
  // its session reading "working" forever — and so never reached Ready for
  // review, which is exactly where stranded work belongs.
  it("is 'idle' when the host has gone offline, even with paneBusy true", () => {
    const s = session({ session: signals({ paneBusy: true, transcriptAgeSec: 5 }) });
    const now = 1_000_000;
    expect(liveState(s, now - 10_000, now)).toBe("working");
    expect(liveState(s, now - 600_000, now)).toBe("idle");
  });

  it("is 'idle' before a transcript exists, whatever paneBusy says", () => {
    const s = session({ session: signals({ paneBusy: true, transcriptAgeSec: null }) });
    expect(liveState(s, Date.now(), Date.now())).toBe("idle");
  });

  it("keeps the old behaviour when no host is supplied", () => {
    const s = session({ session: signals({ paneBusy: true, transcriptAgeSec: 5 }) });
    expect(liveState(s)).toBe("working");
  });

  // XERK-245: a session that delegated work and ended its own turn paints no
  // interrupt hint, so paneBusy reads false while a background agent is still
  // running — and the glyph said idle for work in progress.
  it("is 'working' while background agents run, even with paneBusy false", () => {
    const s = session({ session: signals({
      paneBusy: false, transcriptAgeSec: 999,
      agents: [{ type: "qa", label: "QA the parity change" }],
    }) });
    const now = 1_000_000;
    expect(liveState(s, now - 10_000, now)).toBe("working");
    // An empty list is "no agents", not "can't tell"; an older agent sends none.
    expect(liveState(session({ session: signals({ paneBusy: false, transcriptAgeSec: 999, agents: [] }) }),
      now - 10_000, now)).toBe("idle");
    expect(liveState(session({ session: signals({ paneBusy: false, transcriptAgeSec: 999 }) }),
      now - 10_000, now)).toBe("idle");
    // Behind the offline gate, like paneBusy.
    expect(liveState(s, now - 600_000, now)).toBe("idle");
  });

  it("is 'error' when status is error, regardless of session signals", () => {
    const s = session({ status: "error", session: signals({ question: "pick one" }) });
    expect(liveState(s)).toBe("error");
  });

  it("is 'stopped' when status is stopped, even over a lingering question", () => {
    const s = session({ status: "stopped", session: signals({ question: "pick one" }) });
    expect(liveState(s)).toBe("stopped");
  });

  it("is 'waiting' when a question is pending, even while transcript is fresh", () => {
    const s = session({
      status: "running",
      session: signals({ question: "pick one", transcriptAgeSec: 1 }),
    });
    expect(liveState(s)).toBe("waiting");
  });

  it("is 'working' when paneBusy is true, even with a stale transcript", () => {
    const s = session({
      status: "running",
      session: signals({ paneBusy: true, transcriptAgeSec: 999 }),
    });
    expect(liveState(s)).toBe("working");
  });

  it("is 'idle' when paneBusy is false, even with a fresh transcript", () => {
    const s = session({
      status: "running",
      session: signals({ paneBusy: false, transcriptAgeSec: 1 }),
    });
    expect(liveState(s)).toBe("idle");
  });

  it("still yields to a pending question when paneBusy is true", () => {
    const s = session({
      status: "running",
      session: signals({ paneBusy: true, question: "pick one" }),
    });
    expect(liveState(s)).toBe("waiting");
  });

  it("falls back to transcript freshness when paneBusy is null (older agent)", () => {
    const fresh = session({ status: "running", session: signals({ paneBusy: null, transcriptAgeSec: 89 }) });
    expect(liveState(fresh)).toBe("working");
    const stale = session({ status: "running", session: signals({ paneBusy: null, transcriptAgeSec: 90 }) });
    expect(liveState(stale)).toBe("idle");
  });

  it("is 'working' when the transcript was written to within 90s", () => {
    const s = session({ status: "running", session: signals({ transcriptAgeSec: 89 }) });
    expect(liveState(s)).toBe("working");
  });

  it("is 'idle' when the transcript is stale (>= 90s)", () => {
    const s = session({ status: "running", session: signals({ transcriptAgeSec: 90 }) });
    expect(liveState(s)).toBe("idle");
  });

  it("is 'idle' when running with null session signals", () => {
    const s = session({ status: "running", session: null });
    expect(liveState(s)).toBe("idle");
  });

  it("is 'idle' when transcriptAgeSec is null", () => {
    const s = session({ status: "running", session: signals({ transcriptAgeSec: null }) });
    expect(liveState(s)).toBe("idle");
  });
});

describe("glyph", () => {
  it.each([
    ["working", "!"],
    ["waiting", "?"],
    ["idle", "-"],
    ["stopped", "o"],
    ["error", "x"],
    ["pending", "…"],
  ] as const)("maps %s -> %s", (state, expected) => {
    expect(glyph(state)).toBe(expected);
  });
});

describe("isDsh", () => {
  // The dsh machinery is RETAINED behind the DSH_ENABLED kill switch (currently
  // OFF fleet-wide). These assertions exercise the marker's own logic, so they
  // flip the switch on for the block and reset it after — no leakage into the
  // rest of the suite, which must see the shipped (disabled) behavior.
  beforeEach(() => __setDshEnabled(true));
  afterEach(() => __setDshEnabled(false));

  // XERK-460: "dsh" is the only truthy runtime; the default and every pre-dsh
  // session (claude / "" / absent) must read false so it shows no marker.
  it("is true only for agentType === 'dsh'", () => {
    expect(isDsh(session({ agentType: "dsh" }))).toBe(true);
  });
  it.each([
    ["claude", { agentType: "claude" }],
    ["empty string (pre-dsh agent)", { agentType: "" }],
    ["absent", {}],
    ["null", { agentType: null }],
    ["unknown value", { agentType: "something-else" }],
  ] as const)("is false for %s", (_label, overrides) => {
    expect(isDsh(session(overrides))).toBe(false);
  });

  // With the kill switch OFF (the shipped default) a dsh session is invisible as
  // a dsh session — the marker reads false even for agentType === "dsh".
  it("is false for a dsh session when DSH_ENABLED is off", () => {
    __setDshEnabled(false);
    expect(isDsh(session({ agentType: "dsh" }))).toBe(false);
  });
});

describe("flattenSessions", () => {
  function agent(overrides: Partial<AgentInfo> = {}): AgentInfo {
    return {
      key: "host-a",
      device: "host-a",
      online: true,
      repos: [],
      sessions: [],
      closedSessions: [],
      ...overrides,
    };
  }

  it("sorts hosts by device name and sessions by createdAt within a host", () => {
    const agents: AgentInfo[] = [
      agent({
        key: "zeta",
        device: "zeta",
        sessions: [
          session({ id: "z2", createdAt: "2026-01-02T00:00:00Z" }),
          session({ id: "z1", createdAt: "2026-01-01T00:00:00Z" }),
        ],
      }),
      agent({
        key: "alpha",
        device: "alpha",
        sessions: [session({ id: "a1", createdAt: "2026-01-01T00:00:00Z" })],
      }),
    ];

    const flat = flattenSessions(agents);
    expect(flat.map((r) => `${r.device}:${r.session.id}`)).toEqual([
      "alpha:a1",
      "zeta:z1",
      "zeta:z2",
    ]);
  });

  it("carries the host's online flag onto each session ref", () => {
    const agents: AgentInfo[] = [
      agent({ key: "offhost", device: "offhost", online: false, sessions: [session({ id: "s1" })] }),
    ];
    const flat = flattenSessions(agents);
    expect(flat).toEqual([
      { hostKey: "offhost", device: "offhost", online: false, session: agents[0]!.sessions[0] },
    ]);
  });

  it("returns an empty array for no agents", () => {
    expect(flattenSessions([])).toEqual([]);
  });

  it("skips hosts with no sessions but keeps other hosts' sessions", () => {
    const agents: AgentInfo[] = [
      agent({ key: "empty", device: "empty", sessions: [] }),
      agent({ key: "busy", device: "busy", sessions: [session({ id: "b1" })] }),
    ];
    expect(flattenSessions(agents).map((r) => r.session.id)).toEqual(["b1"]);
  });
});

File: android/app/src/test/java/com/xerktech/turma/core/SessionsTest.kt
package com.xerktech.turma.core

import com.xerktech.turma.model.AgentInfo
import com.xerktech.turma.model.GitState
import com.xerktech.turma.model.LiveSignals
import com.xerktech.turma.model.SessionInfo
import org.junit.Assert.assertEquals
import org.junit.Test

class SessionsTest {

    private val now = 1_000_000L

    @Test fun `paneBusy true means working regardless of freshness`() {
        val s = SessionInfo(status = "running", session = LiveSignals(paneBusy = true, transcriptAgeSec = 9999.0))
        assertEquals(LiveState.WORKING, liveState(s, now, now))
    }

    @Test fun `paneBusy false means idle even with fresh transcript`() {
        val s = SessionInfo(status = "running", session = LiveSignals(paneBusy = false, transcriptAgeSec = 1.0))
        assertEquals(LiveState.IDLE, liveState(s, now, now))
    }

    // XERK-245: a session that delegated work and ended its own turn paints no
    // interrupt hint, so paneBusy reads false while a background agent is still
    // running — and the card said idle, and it qualified as Ready for review,
    // buzzing the phone mid-run.
    @Test fun `live background agents mean working even with paneBusy false`() {
        val bg = LiveSignals(
            paneBusy = false, transcriptAgeSec = 999.0,
            agents = listOf(com.xerktech.turma.model.LiveAgent(type = "qa", label = "QA it")),
        )
        val s = SessionInfo(status = "running", session = bg)
        assertEquals(LiveState.WORKING, liveState(s, now, now))
        // Behind the offline gate, like paneBusy: a host that died mid-run must
        // not leave its sessions reading working forever.
        assertEquals(LiveState.IDLE, liveState(s, now - 120_000L, now))
        // An empty list is "no agents"; an older agent reports none at all.
        assertEquals(
            LiveState.IDLE,
            liveState(SessionInfo(status = "running",
                session = LiveSignals(paneBusy = false, transcriptAgeSec = 999.0)), now, now),
        )
    }

    // The card's wording, which no test covered — a QA mutation pass removed the
    // background-agent branch of liveStateLabel and every gate stayed green.
    @Test fun `liveStateLabel names background agents and pluralizes them`() {
        val one = LiveSignals(agents = listOf(com.xerktech.turma.model.LiveAgent(type = "Explore")))
        val two = LiveSignals(agents = listOf(
            com.xerktech.turma.model.LiveAgent(type = "Explore"),
            com.xerktech.turma.model.LiveAgent(type = "general-purpose"),
        ))
        assertEquals("1 background agent",
            com.xerktech.turma.ui.liveStateLabel(LiveState.WORKING, one))
        assertEquals("2 background agents",
            com.xerktech.turma.ui.liveStateLabel(LiveState.WORKING, two))
        // No agents, or a state that isn't WORKING: the plain state word.
        assertEquals("working",
            com.xerktech.turma.ui.liveStateLabel(LiveState.WORKING, LiveSignals()))
        assertEquals("working", com.xerktech.turma.ui.liveStateLabel(LiveState.WORKING, null))
        assertEquals("idle", com.xerktech.turma.ui.liveStateLabel(LiveState.IDLE, one))
        assertEquals("waiting", com.xerktech.turma.ui.liveStateLabel(LiveState.WAITING, one))
    }

    // XERK-538: a QA / QA-delta pass reads "QA Review" while the card stays
    // WORKING (Active) — matching the web's qaReviewing/agentWorkLabel.
    @Test fun `liveStateLabel reads QA Review for a QA agent`() {
        val qa = LiveSignals(agents = listOf(com.xerktech.turma.model.LiveAgent(type = "qa")))
        // qa-delta wins even beside an ordinary agent.
        val delta = LiveSignals(agents = listOf(
            com.xerktech.turma.model.LiveAgent(type = "Explore"),
            com.xerktech.turma.model.LiveAgent(type = "qa-delta"),
        ))
        assertEquals("QA Review", com.xerktech.turma.ui.liveStateLabel(LiveState.WORKING, qa))
        assertEquals("QA Review", com.xerktech.turma.ui.liveStateLabel(LiveState.WORKING, delta))
        // Only while WORKING — an idle/waiting card keeps its plain word.
        assertEquals("idle", com.xerktech.turma.ui.liveStateLabel(LiveState.IDLE, qa))
    }

    @Test fun `a session waiting on background agents is not ready for review`() {
        val bg = SessionInfo(
            status = "running",
            session = LiveSignals(
                paneBusy = false, transcriptAgeSec = 5.0, lastRole = "assistant",
                agents = listOf(com.xerktech.turma.model.LiveAgent(type = "qa", label = "QA it")),
            ),
        )
        assertEquals(false, readyForReview(bg, liveState(bg, now, now)))
        // Once the agent finishes the list empties and it qualifies as before.
        val done = bg.copy(session = bg.session!!.copy(agents = emptyList()))
        assertEquals(true, readyForReview(done, liveState(done, now, now)))
    }

    // ---- readyForReview (XERK-224) ------------------------------------------
    // The port of the web's rule (turma/public/sessions.html), which the hub's
    // ready-for-review alert mirrors again — all three decide the same group.

    private fun quiet(
        lastRole: String = "",
        prs: List<com.xerktech.turma.model.PrInfo> = emptyList(),
        newWork: Boolean = false,
    ) = SessionInfo(
        status = "running", prs = prs, newWorkSincePrs = newWork,
        session = LiveSignals(paneBusy = false, lastRole = lastRole),
    )
    private fun pr(state: String) = com.xerktech.turma.model.PrInfo(url = "u$state", state = state)

    @Test fun `readyForReview takes a question, an unlanded PR, or a finished turn`() {
        assertEquals(true, readyForReview(quiet(), LiveState.WAITING))
        assertEquals(false, readyForReview(quiet(), LiveState.WORKING))
        assertEquals(false, readyForReview(quiet(), LiveState.STOPPED))
        // The no-PR research task — the case a PR-only rule would miss.
        assertEquals(true, readyForReview(quiet(lastRole = "assistant"), LiveState.IDLE))
        // Nothing written yet, or the last word was the operator's: not review-ready.
        assertEquals(false, readyForReview(quiet(), LiveState.IDLE))
        assertEquals(false, readyForReview(quiet(lastRole = "user"), LiveState.IDLE))
    }

    @Test fun `readyForReview judges a PR-bearing session on the PR alone`() {
        assertEquals(true, readyForReview(quiet(prs = listOf(pr("OPEN"))), LiveState.IDLE))
        assertEquals(true, readyForReview(quiet(prs = listOf(pr("DRAFT"))), LiveState.IDLE))
        // An unfetched state must never be what drops work off the list.
        assertEquals(true, readyForReview(quiet(prs = listOf(pr(""))), LiveState.IDLE))
        // Merging (or closing) IS the review — park it in Idle until verified.
        assertEquals(false, readyForReview(quiet(prs = listOf(pr("MERGED"))), LiveState.IDLE))
        assertEquals(false, readyForReview(quiet(prs = listOf(pr("CLOSED"))), LiveState.IDLE))
        assertEquals(true, readyForReview(quiet(prs = listOf(pr("MERGED"), pr("OPEN"))), LiveState.IDLE))
        // A landed PR outranks the finished turn that opened it, else merging
        // could never move a session out of the section.
        assertEquals(false, readyForReview(quiet("assistant", listOf(pr("MERGED"))), LiveState.IDLE))
    }

    @Test fun `readyForReview lets a merged PR stop hiding the session's next task`() {
        // A session is a conversation, not a pull request: hand the same one a
        // new task after the merge and the PR it already shipped must not bury
        // the result (XERK-224).
        val merged = listOf(pr("MERGED"))
        assertEquals(false, readyForReview(quiet("assistant", merged), LiveState.IDLE))
        assertEquals(
            "new work after the merge is still work awaiting review",
            true,
            readyForReview(quiet("assistant", merged, newWork = true), LiveState.IDLE),
        )
        // The expiry only lifts the demotion — it is not itself a qualifier.
        assertEquals(false, readyForReview(quiet("user", merged, newWork = true), LiveState.IDLE))
        // And an unlanded PR is decided before the question is even asked.
        assertEquals(true, readyForReview(quiet("user", listOf(pr("MERGED"), pr("OPEN"))), LiveState.IDLE))
    }

    @Test fun `falls back to transcript freshness when paneBusy unknown`() {
        val fresh = SessionInfo(status = "running", session = LiveSignals(paneBusy = null, transcriptAgeSec = 5.0))
        assertEquals(LiveState.WORKING, liveState(fresh, now, now))
        val stale = SessionInfo(status = "running", session = LiveSignals(paneBusy = null, transcriptAgeSec = 200.0))
        assertEquals(LiveState.IDLE, liveState(stale, now, now))
    }

    @Test fun `a pending question wins over working`() {
        val s = SessionInfo(status = "running", session = LiveSignals(paneBusy = true, question = "Proceed?"))
        assertEquals(LiveState.WAITING, liveState(s, now, now))
    }

    @Test fun `stopped session is STOPPED`() {
        assertEquals(LiveState.STOPPED, liveState(SessionInfo(status = "stopped"), now, now))
    }

    @Test fun `branch shows detached until the agent branches`() {
        assertEquals("detached", sessionBranch(SessionInfo(git = GitState(branch = "HEAD"))))
        assertEquals("feat/x", sessionBranch(SessionInfo(git = GitState(branch = "feat/x"))))
    }

    @Test fun `name prefers summary then label then worktree`() {
        assertEquals("Fix login", sessionName(SessionInfo(summary = "Fix login", label = "l", worktreePath = "/a/b")))
        assertEquals("mylabel", sessionName(SessionInfo(label = "mylabel", worktreePath = "/a/wt-9")))
        assertEquals("wt-9", sessionName(SessionInfo(worktreePath = "/a/wt-9")))
    }

    @Test fun `card repo label names the repo, or says repos root in words`() {
        assertEquals("Turma", sessionRepoLabel(SessionInfo(repo = "Turma")))
        // A repos-root session: by the flag, and by the agent's "(root)" sentinel
        // (a closed/resumable record can carry the name without the flag).
        assertEquals("repos root", sessionRepoLabel(SessionInfo(repo = "(root)", root = true)))
        assertEquals("repos root", sessionRepoLabel(SessionInfo(repo = "(root)")))
        // Nothing reported at all — never blank, so the row can't lose a separator.
        assertEquals("?", sessionRepoLabel(SessionInfo(repo = "")))
    }

    @Test fun `header meta joins host repo and branch, dropping blanks`() {
        assertEquals(
            "truenas · Turma · XERK-121",
            sessionHeaderMeta("truenas", SessionInfo(repo = "Turma", git = GitState(branch = "XERK-121"))),
        )
        // No repo (repos-root) or no branch (detached) still reads cleanly.
        assertEquals("truenas · detached", sessionHeaderMeta("truenas", SessionInfo(repo = "")))
        assertEquals(
            "truenas · Turma · detached",
            sessionHeaderMeta("truenas", SessionInfo(repo = "Turma", git = GitState(branch = "HEAD"))),
        )
    }

    // ---- liveMarker / tunnelOnlineOf (web's tunnel chip, XERK-252) -----------

    @Test fun `a host with no tunnel outranks our own open socket`() {
        // The hub holds our /live socket across a control-channel flap, so
        // `connected` stays true while nothing flows. Saying "live" there claims
        // a stream that has stopped.
        assertEquals("⚠ tunnel offline", liveMarker(tunnelOnline = false, connected = true))
        assertEquals("⚠ tunnel offline", liveMarker(tunnelOnline = false, connected = false))
    }

    @Test fun `a healthy host marks live only while our socket is up`() {
        assertEquals("live", liveMarker(tunnelOnline = true, connected = true))
        assertEquals("", liveMarker(tunnelOnline = true, connected = false))
    }

    @Test fun `a host missing from the payload is unknown, never offline`() {
        // Nothing heard about the host is not a fault we may claim (and it is
        // the state every screen starts in, before the first fleet payload).
        assertEquals(true, tunnelOnlineOf(null))
        assertEquals(true, tunnelOnlineOf(AgentInfo(key = "h1", terminalOnline = true)))
        assertEquals(false, tunnelOnlineOf(AgentInfo(key = "h1", terminalOnline = false)))
    }

    // ---- workLine (web index.html workLine/unpushedCommits, XERK-78) ---------

    private fun sessWork(
        pushed: Boolean? = null, aheadOfBase: Int? = null, aheadOfRemote: Int? = null,
        baseRef: String? = null, dirty: Int = 0,
    ) = SessionInfo(
        work = com.xerktech.turma.model.WorkInfo(
            baseRef = baseRef, aheadOfBase = aheadOfBase, pushed = pushed, aheadOfRemote = aheadOfRemote,
        ),
        git = GitState(dirtyFiles = dirty),
    )

    @Test fun `workLine is null when nothing is known`() {
        assertEquals(null, workLine(SessionInfo()))
        assertEquals(null, workLine(sessWork()))
    }

    @Test fun `unpushed commits or dirty files read as risk`() {
        val risky = workLine(sessWork(pushed = false, aheadOfBase = 3, baseRef = "main", dirty = 2))!!
        assertEquals("3 commits ahead of main · not pushed · 2 dirty files", risky.text)
        assertEquals(true, risky.risk)
        // Singulars singular.
        val one = workLine(sessWork(pushed = false, aheadOfBase = 1, dirty = 1))!!
        assertEquals("1 commit ahead · not pushed · 1 dirty file", one.text)
    }

    @Test fun `pushed and clean reads safe`() {
        val safe = workLine(sessWork(pushed = true, aheadOfBase = 2, aheadOfRemote = 0, baseRef = "main"))!!
        assertEquals("2 commits ahead of main · pushed", safe.text)
        assertEquals(false, safe.risk)
        // Pushed with unknown sync says so rather than claiming either way.
        val unknown = workLine(sessWork(pushed = true, aheadOfBase = 2, aheadOfRemote = null))!!
        assertEquals("2 commits ahead · pushed · sync unknown", unknown.text)
        assertEquals(false, unknown.risk)
        // Pushed but with commits origin doesn't have yet: risk again.
        val behind = workLine(sessWork(pushed = true, aheadOfBase = 5, aheadOfRemote = 2))!!
        assertEquals("5 commits ahead · 2 unpushed", behind.text)
        assertEquals(true, behind.risk)
    }

    // --- eligibleMoveTargets (XERK-101) --------------------------------------

    private fun agent(
        key: String,
        online: Boolean = true,
        org: String = "org.a",
        repos: List<String> = listOf("repoX"),
        sessions: List<SessionInfo> = emptyList(),
    ) = com.xerktech.turma.model.AgentInfo(
        key = key, device = key, online = online,
        jira = com.xerktech.turma.model.JiraBlock(siteKey = org),
        repos = repos.map { com.xerktech.turma.model.RepoInfo(name = it) },
        sessions = sessions,
    )

    @Test fun `move targets are online same-org hosts with the repo, minus the source`() {
        val sess = SessionInfo(id = "s1", status = "running", repo = "repoX")
        val agents = listOf(
            agent("src", sessions = listOf(sess)),
            agent("ok"),                              // eligible
            agent("off", online = false),             // offline
            agent("otherOrg", org = "org.b"),         // different org
            agent("noRepo", repos = listOf("other")), // lacks the repo
        )
        val targets = eligibleMoveTargets(agents, "src", sess).map { it.key }
        assertEquals(listOf("ok"), targets)
    }

    @Test fun `no eligible targets when the org has only the source host`() {
        val sess = SessionInfo(id = "s1", status = "running", repo = "repoX")
        val agents = listOf(agent("src", sessions = listOf(sess)))
        assertEquals(emptyList<String>(), eligibleMoveTargets(agents, "src", sess).map { it.key })
    }
}
